src/core/cycle_detector.go

```diff
@@ -5,108 +5,60 @@ import (
 	"strings"
 )
 
-type dependencyChain []*BuildLabel
-type dependencyLink struct {
-	from *BuildLabel
-	to   *BuildLabel
-}
-
-func (c dependencyChain) String() string {
-	labels := make([]string, len(c))
-	for i, l := range c {
-		labels[i] = l.String()
-	}
-	return strings.Join(labels, "\n -> ")
-}
-
 type cycleDetector struct {
-	deps  map[*BuildLabel][]*BuildLabel
-	queue chan dependencyLink
-}
-
-func newCycleDetector() *cycleDetector {
-	c := new(cycleDetector)
-
-	// Set this to an order of magnitude that seems sensible to avoid resizing them rapidly initially
-	c.deps = make(map[*BuildLabel][]*BuildLabel, 1000)
-
-	// Buffer this a little so handle spikes in targets coming in
-	c.queue = make(chan dependencyLink, 1000)
-
-	return c
+	graph *BuildGraph
 }
 
-// AddDependency queues up another dependency for the cycle detector to check
-func (c *cycleDetector) AddDependency(from *BuildLabel, to *BuildLabel) {
-	go func() {
-		c.queue <- dependencyLink{from: from, to: to}
-	}()
-}
-
-// checkForCycle just checks to see if there's a dependency cycle. It doesn't compute the cycle to avoid excess
-// allocations. buildCycle can be used to reconstruct the cycle once one has been found.
-func (c *cycleDetector) checkForCycle(head, tail *BuildLabel, done map[BuildLabel]struct{}) bool {
-	if _, ok := done[*tail]; ok {
-		return false
-	}
-	done[*tail] = struct{}{}
-	if tailDeps, ok := c.deps[tail]; ok {
-		for _, dep := range tailDeps {
-			if dep == head {
-				// If the tail has a dependency on the head, we've found a cycle
-				return true
-			}
-
-			if c.checkForCycle(head, dep, done) {
-				return true
+// Check runs a single check of the build graph to see if any cycles can be detected.
+// If it finds one an errCycle is returned.
+func (c *cycleDetector) Check() *errCycle {
+	log.Debug("Running cycle detection...")
+	permanent := map[*BuildTarget]struct{}{}
+	temporary := map[*BuildTarget]struct{}{}
+
+	var visit func(target *BuildTarget) ([]*BuildTarget, bool)
+	visit = func(target *BuildTarget) ([]*BuildTarget, bool) {
+		if _, present := permanent[target]; present {
+			return nil, false
+		} else if _, present := temporary[target]; present {
+			return []*BuildTarget{target}, false
+		}
```

Comment:
Is this how we ensure we get the shorter chain? Smart. Would be nice if we could build the chain the other way because appending to a slice is cheaper but I don't have any ideas.

Reply:
Yeah there's a bit of faffing with the second return value (which is basically to chop it off so a -> b -> c -> d -> b doesn't return a as well).

Good point, I thought the same on the slices, but at the point we're building up the slice we're doomed anyway so I don't think it will matter much.

Reply:
Ah oh yeah, we only return the chain if we find something! That's actually really clever.